File: Assets/HealthBar.js
#pragma strict

public var maxLength : float = 0;
public var gameManager : GameManager;
public var currentPercentage : float = 0;
public var targetPercentage : float = 0;

function SetColor(color : Color) {
	renderer.material.color = color;
}

function SetMeshBounds(l : float, h:float) {
	var mesh = GetComponent(MeshFilter).mesh;
	var vertices : Vector3[] = mesh.vertices;
	vertices[0] = Vector3(0,-0.5*h,0);
	vertices[1] = Vector3(l,0.5*h,0);
	vertices[2] = Vector3(l,-0.5*h,0);
	vertices[3] = Vector3(0,0.5*h,0);
	mesh.vertices = vertices;
	mesh.RecalculateBounds();
	
	var s = 1.0;
	renderer.material.mainTextureScale = Vector2(l*s,1.0*s);
}

function GetHeight() {
	return maxLength*0.08f;
}

function IsEmpty() {
	return targetPercentage == 0;
}

function IsRed() {
	return targetPercentage < 0.33f;
}

function IsYellow() {
	return targetPercentage < 0.66f;
}

function IsGreen() {
	return targetPercentage >= 0.66f;
}

function SetProgress(p : float) {
	currentPercentage = p;
	SetMeshBounds(maxLength * p, GetHeight());
	
	if (IsRed()) {
		SetColor(Color.red);
	} else if (IsYellow()) {
		SetColor(Color.yellow);
	} else {
		SetColor(Color.green);
	}
}

function SetPercentage(p : float) {
	Debug.Log("set healthbar pct = " + p);
	targetPercentage = p;
	var endPercentage = targetPercentage;
	var startPercentage = currentPercentage;
	var duration = 0.5f;
	var rate = 1.0/duration;
	var t = 0.0f;
	while (t < 1.0) {
		if (endPercentage != targetPercentage) {
			// pct changed during animation, cancel it
			return;
		}
		t += Time.deltaTime * rate;
		SetProgress(startPercentage + (endPercentage-startPercentage)*t);
		yield;
	}
}

function Start () {
	maxLength = 0;
}

function Update () {
}

File: Assets/cs/HealthBar.cs
using UnityEngine;
using System;
using UnityEngine.UI;
using System.Collections;


public class HealthBar:MonoBehaviour{
	
	public GameManager gameManager;
	public float currentPercentage = 0.0f;
	public float targetPercentage = 0.0f;
	public float maxLength;
	public Text healthLabel;
	
	public RectTransform rectTransform;
	public Image image;
	
	public bool IsEmpty() {
		return targetPercentage == 0;
	}
	
	public bool IsRed() {
		return targetPercentage < 0.50f;
	}
	
	public bool IsYellow() {
		return targetPercentage < 1.00f;
	}
	
	public bool IsGreen()  {
		return targetPercentage >= 1.00f;
	}
	
	public void SetColor(Color color) {
		image.color = color;
	}
	
	public float GetHeight() {
		return (float)rectTransform.sizeDelta[1];
	}
	
	public void SetProgress(float p) {
		currentPercentage = p;
		healthLabel.text = String.Format("{0}x", Mathf.RoundToInt(p*100.0f));
		if (p > 1.0f) {
			p = 1.0f;
		}
		Vector2 newSize = new Vector2(maxLength*p, GetHeight());
		rectTransform.sizeDelta = newSize;
		
		if (IsRed()) {
			SetColor(Color.red);
		} else if (IsYellow()) {
			SetColor(Color.yellow);
		} else {
			SetColor(Color.green);
		}
	}
	
	public IEnumerator SetPercentage(float p) {
		//Debug.Log("set healthbar pct = " + p);
		targetPercentage = p;
		float endPercentage = targetPercentage;
		float startPercentage = currentPercentage;
		
		// make sure percentage is not below 0.05f for graphic purposes
		if (startPercentage < 0.05f) {
			startPercentage = 0.05f;
		}
		
		if (endPercentage == 0) {
			if (startPercentage <= 0.05f) {
				startPercentage = 0.0f;
			}
		}
		float duration = 1.0f;
		float rate = 1.0f/duration;
		float t = 0.0f;
		while (t < 1.0f) {
			if (endPercentage != targetPercentage) {
				// pct changed during animation, cancel it
				break;
			}
			t += Time.deltaTime * rate;
			SetProgress(startPercentage + (endPercentage-startPercentage)*t);
			yield return null;
		}
		SetProgress(endPercentage);
	}
	
	public void Start() {
		maxLength = (float)rectTransform.sizeDelta[0];
	}
	
	public void Update() {
	}
}
